pkgs/by-name/op/openai-triton-llvm/package.nix

```diff
@@ -69,57 +93,83 @@ stdenv.mkDerivation (finalAttrs: {
   sourceRoot = "${finalAttrs.src.name}/llvm";
 
   cmakeFlags = [
-    "-DLLVM_TARGETS_TO_BUILD=${
-      let
-        # Targets can be found in
-        # https://github.com/llvm/llvm-project/tree/f28c006a5895fc0e329fe15fead81e37457cb1d1/clang/lib/Basic/Targets
-        # NOTE: Unsure of how "host" would function, especially given that we might be cross-compiling.
-        llvmTargets = [ "AMDGPU" "NVPTX" ]
-        ++ lib.optionals stdenv.isAarch64 [ "AArch64" ]
-        ++ lib.optionals stdenv.isx86_64 [ "X86" ];
-      in
-      lib.concatStringsSep ";" llvmTargets
-    }"
-    "-DLLVM_ENABLE_PROJECTS=llvm;mlir"
-    "-DLLVM_INSTALL_UTILS=ON"
-  ] ++ lib.optionals (buildDocs || buildMan) [
-    "-DLLVM_INCLUDE_DOCS=ON"
-    "-DMLIR_INCLUDE_DOCS=ON"
-    "-DLLVM_BUILD_DOCS=ON"
-    # "-DLLVM_ENABLE_DOXYGEN=ON" Way too slow, only uses one core
-    "-DLLVM_ENABLE_SPHINX=ON"
-    "-DSPHINX_OUTPUT_HTML=ON"
-    "-DSPHINX_OUTPUT_MAN=ON"
-    "-DSPHINX_WARNINGS_AS_ERRORS=OFF"
-  ] ++ lib.optionals buildTests [
-    "-DLLVM_INCLUDE_TESTS=ON"
-    "-DMLIR_INCLUDE_TESTS=ON"
-    "-DLLVM_BUILD_TESTS=ON"
-  ];
+    (lib.cmakeFeature "LLVM_TARGETS_TO_BUILD" (lib.concatStringsSep ";" llvmTargetsToBuild'))
+    (lib.cmakeFeature "LLVM_ENABLE_PROJECTS" (lib.concatStringsSep ";" llvmProjectsToBuild))
+    (lib.cmakeFeature "LLVM_HOST_TRIPLE" stdenv.hostPlatform.config)
+    (lib.cmakeFeature "LLVM_DEFAULT_TARGET_TRIPLE" stdenv.hostPlatform.config)
+    (lib.cmakeBool "LLVM_INSTALL_UTILS" true)
+    (lib.cmakeBool "LLVM_INCLUDE_DOCS" (buildDocs || buildMan))
+    (lib.cmakeBool "MLIR_INCLUDE_DOCS" (buildDocs || buildMan))
+    (lib.cmakeBool "LLVM_BUILD_DOCS" (buildDocs || buildMan))
+    # Way too slow, only uses one core
+    # (lib.cmakeBool "LLVM_ENABLE_DOXYGEN" (buildDocs || buildMan))
+    (lib.cmakeBool "LLVM_ENABLE_SPHINX" (buildDocs || buildMan))
+    (lib.cmakeBool "SPHINX_OUTPUT_HTML" buildDocs)
+    (lib.cmakeBool "SPHINX_OUTPUT_MAN" buildMan)
+    (lib.cmakeBool "SPHINX_WARNINGS_AS_ERRORS" false)
+    (lib.cmakeBool "LLVM_INCLUDE_TESTS" buildTests)
+    (lib.cmakeBool "MLIR_INCLUDE_TESTS" buildTests)
+    (lib.cmakeBool "LLVM_BUILD_TESTS" buildTests)
+  ] ++ lib.optionals (!isNative) (let
+    nativeToolchainFlags = let
+      nativeCC = pkgsBuildBuild.targetPackages.stdenv.cc;
```

Comment:
Oh... could you briefly explain what is the difference between `pkgsBuildBuild.pkgsTargetTarget` and `pkgsBuildHost`? If I were trying to implement the cross support myself, how would I figure out I need to use this particular incantation?

Reply:
If you meant ``pkgsBuildBuild`` and ``pkgsTargetTarget`` separately, then I believe ``pkgsBuildBuild`` is just the ``pkgs`` that the builder is using, and ``pkgsTargetTarget`` being the build target's ``pkgs``.
I'm not sure what the difference between ``pkgsTargetTarget`` and ``pkgsBuildHost`` would be though.
Or for that matter, with the two above, what the difference with ``pkgsHostHost`` would be.

Reply:
Would ``pkgsBuildHost`` be the builder's target's ``pkgs``?
So confusing...

Reply:
From what I can tell, ``pkgsTargetTarget`` is basically ``pkgsHostHost`` but for runtime or specialized stuff? Like a compiler?
I could be wrong, but you pretty much never want to use ``pkgsTargetTarget`` over ``pkgsHostHost``.
I don't fully understand it myself, but see: https://github.com/NixOS/nixpkgs/blob/master/doc/stdenv/cross-compilation.chapter.md

Reply:
Specifically, ``pkgsTargetTarget`` seems to be used when the architecture can't be hot-plugged at runtime and you need to specify variations ahead of time in the derivation.
I think that should more or less resolve this question.

Reply:
I somehow managed to not even look at the code this comment is on.
Okay, I think I understand why you asked this now.
The last comment still stands, and AFAIK this (``pkgsBuildBuild.targetPackages``) should be correct usage.
I wonder if it might be a legacy thing, but shouldn't ``pkgsBuildTarget`` work just as well?